feat: add pull flow testing (#504)
FilOzone/dealbot: docs/checks/events-and-metrics.md

Reviewer: I suggested some new events and metric names when looking holistically at the terms for the pull check flow. 

I leaned into the idea of there is a pull requst that is submitted, acknowledged, started, and finished.

Whatever event/metric names we decide will need to be propagated through the rest of the doc and in code.

Author: I've updated names to suggested names for events and metrics in [445a0d6](https://github.com/FilOzone/dealbot/pull/504/commits/445a0d6c63b5b215fbcb50d6f3c5615fa471bff9)

## Pull Check Event Model

Below are the events for a [Pull Check](./pull-check.md). Pull checks reverse the data flow of the [Data Storage check](./data-storage.md): instead of dealbot uploading bytes, it asks the SP to pull bytes from a temporary `/api/piece/{pieceCid}` URL.

### Pull Check Event Timeline

```mermaid
sequenceDiagram
  autonumber
  participant Dealbot
  participant SP as PDP Storage Provider
  participant RPC as Chain RPC Provider

  Dealbot->>Dealbot: hostedPieceRegistered
  Dealbot->>SP: pullRequestSubmitted (pullPieces)
  SP-->>Dealbot: pullRequestAcknowledged
  SP-->>Dealbot: hostedPieceFirstByteRead
  Dealbot->>SP: pullStatusPolled (waitForPullStatus, repeated)
  SP-->>Dealbot: pullTerminalStatusReported
  Dealbot->>SP: directPieceFetchStarted (/piece/{cid})
  SP-->>Dealbot: directPieceFetchCompleted
  Dealbot-->>Dealbot: pullCheckIntegrityChecked
```

### Pull Check Event List

| Event | Definition | Implemented | Source of truth |
|------|------------|:------:|-----------------|

> | <a id="pullRequestSubmitted"></a>`pullRequestSubmitted` | Dealbot calls `pullPieces` against the SP for the registered piece CID. | Yes | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |

Reviewer: ```suggestion
| <a id="pullRequestSubmittedToSp"></a>`pullRequestSubmittedToSp` | Dealbot calls Synpase pullPiece (`POST /pdp/piece/pull`) against the SP for the registered piece CID. | Yes | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |
```

> | <a id="pullRequestAcknowledged"></a>`pullRequestAcknowledged` | SP returns from `pullPieces` (success or non-terminal-failure). | Yes | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |

Reviewer: ```suggestion
| <a id="pullRequestAcknowledgedBySp"></a>`pullRequestAcknowledgedBySp` | SP returns from `pullPieces` (success or non-terminal-failure). | Yes | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |
```

> | <a id="hostedPieceFirstByteRead"></a>`hostedPieceFirstByteRead` | SP reads the first byte of `/api/piece/{pieceCid}` from dealbot. Recorded once per registration. | Yes | [`pull-piece.controller.ts`](../../apps/backend/src/pull-check/pull-piece.controller.ts) |

Reviewer: ```suggestion
| <a id="pullRequestStartedBySp"></a>`pullRequestStartedBySp` | Dealbot receives SP request for `/api/piece/{pieceCid}`. Recorded once per registration. | Yes | [`pull-piece.controller.ts`](../../apps/backend/src/pull-check/pull-piece.controller.ts) |
```

> | <a id="pullTerminalStatusReported"></a>`pullTerminalStatusReported` | SP reports a terminal pull status (`complete`, `failed`, ...) via `waitForPullStatus`. Intermediate poll statuses are not counted. | Yes | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |

Reviewer: ```suggestion
| <a id="pullRequestIsTerminal"></a>`pullRequestIsTerminal` | Dealbot determines the the pull request is in terminal pull status (`complete`, `failed`, ...) via `waitForPullStatus` or the polling operation has timed out. Intermediate poll statuses are not counted. | Yes | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |
```

> | <a id="pullCheckIntegrityChecked"></a>`pullCheckIntegrityChecked` | Direct `/piece/{pieceCid}` fetch from the SP returns bytes whose recomputed pieceCid matches the expected CID. | Yes | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |

Reviewer: ```suggestion
| <a id="pullRequestIntegrityChecked"></a>`pullRequestIntegrityChecked` | Dealbot performs direct `/piece/{pieceCid}` retrieval from the SP and confirms the bytes match the pieceCid. | Yes | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |
```

> | <a id="pullRequestIntegrityChecked"></a>`pullRequestIntegrityChecked` | Dealbot performs direct `/piece/{pieceCid}` retrieval from the SP and confirms the bytes match the pieceCid. | Yes | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |

Reviewer: ```suggestion
| <a id="pullRequestIntegrityChecked"></a>`pullRequestIntegrityChecked` | Dealbot completed direct `/piece/{pieceCid}` retrieval from the SP and confirmed the bytes match the pieceCid. | Yes | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |
```

| <a id="dataStorageCheckMs"></a>`dataStorageCheckMs` | Data Storage | [`uploadToSpStart`](#uploadToSpStart) | [`ipfsRetrievalIntegrityChecked`](#ipfsRetrievalIntegrityChecked) | Duration of a Data Storage check | |
| <a id="retrievalCheckMs"></a>`retrievalCheckMs` | Retrieval | Retrieval check start | [`ipfsRetrievalIntegrityChecked`](#ipfsRetrievalIntegrityChecked) | Duration of a Retrieval check | |
| <a id="dataSetCreationMs"></a>`dataSetCreationMs` | Data-Set Creation | Data-set creation uploadToSpStart | Data-set creation pieceConfirmed | Duration of one data-set creation with confirmed piece (all using `createDataSetWithPiece`) | [`deal.service.ts`](../../apps/backend/src/deal/deal.service.ts) |

> | <a id="pullCheckRequestLatencyMs"></a>`pullCheckRequestLatencyMs` | Pull | [`pullRequestSubmitted`](#pullRequestSubmitted) | [`pullRequestAcknowledged`](#pullRequestAcknowledged) | Time from `pullPieces` submission to SP request acknowledgement. | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |

Reviewer: ```suggestion
| <a id="pullRequestAcknowledgementLatencyMs"></a>`pullRequestAcknowledgementLatencyMs` | Pull | [`pullRequestSubmitted`](#pullRequestSubmitted) | [`pullRequestAcknowledged`](#pullRequestAcknowledged) | Time from `pullPieces` submission to SP request acknowledgement. | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |
```

> | <a id="pullCheckCompletionLatencyMs"></a>`pullCheckCompletionLatencyMs` | Pull | [`pullRequestSubmitted`](#pullRequestSubmitted) | [`pullTerminalStatusReported`](#pullTerminalStatusReported) | Time from `pullPieces` submission to terminal SP pull status. Observed once on success and once on failure. | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |

Reviewer: ```suggestion
| <a id="pullRequestCompletionLatencyMs"></a>`pullRequestCompletionLatencyMs` | Pull | [`pullRequestSubmitted`](#pullRequestSubmitted) | [`pullTerminalStatusReported`](#pullTerminalStatusReported) | Time from `pullPieces` submission to terminal SP pull status. Observed once on success and once on failure. | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |
```

> | <a id="pullCheckFirstByteMs"></a>`pullCheckFirstByteMs` | Pull | [`pullRequestSubmitted`](#pullRequestSubmitted) | [`hostedPieceFirstByteRead`](#hostedPieceFirstByteRead) | Time from `pullPieces` submission to the SP reading the first byte of `/api/piece/{pieceCid}`. Skipped (no observation) when the SP serves the pull from a local cache and never fetches from dealbot. | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts), [`pull-piece.controller.ts`](../../apps/backend/src/pull-check/pull-piece.controller.ts) |

Reviewer: ```suggestion
| <a id="pullRequestStartedMs"></a>`pullRequestStartedMs` | Pull | [`pullRequestSubmitted`](#pullRequestSubmitted) | [`hostedPieceFirstByteRead`](#hostedPieceFirstByteRead) | Time from `pullPieces` submission to the SP reading the first byte of `/api/piece/{pieceCid}`. Skipped (no observation) when the SP serves the pull from a local cache and never fetches from dealbot. | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts), [`pull-piece.controller.ts`](../../apps/backend/src/pull-check/pull-piece.controller.ts) |
```

We should move this up a line since it happens earlier chronologicially.

> | <a id="pullRequestCompletionLatencyMs"></a>`pullRequestCompletionLatencyMs` | Pull | [`pullRequestSubmittedToSp`](#pullRequestSubmittedToSp) | [`pullRequestIsTerminal`](#pullRequestIsTerminal) | Time from `pullPieces` submission to terminal SP pull status. Observed once on success and once on failure. | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |

Reviewer: ```suggestion
| <a id="pullRequestCompletionLatencyMs"></a>`pullRequestCompletionLatencyMs` | Pull | [`pullRequestSubmittedToSp`](#pullRequestSubmittedToSp) | [`pullRequestIsTerminal`](#pullRequestIsTerminal) | Time from `pullPieces` submission to terminal SP pull status. Emitted once for the check, either on success or failure. | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |
```

Tying to make clearer that there is only one data point emitted here.

> | <a id="pullCheckThroughputBps"></a>`pullCheckThroughputBps` | Pull | n/a | n/a | `(pieceSizeBytes / pullCheckCompletionLatencyMs) * 1000`. Upper-bound on actual transfer rate because `pullCheckCompletionLatencyMs` includes SP-side scheduling and dealbot's polling cadence. | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |

Reviewer: ```suggestion
| <a id="pullRequestThroughputBps"></a>`pullRequestThroughputBps` | Pull | n/a | n/a | `(pieceSizeBytes / pullCheckCompletionLatencyMs) * 1000`. Upper-bound on actual transfer rate because `pullCheckCompletionLatencyMs` includes SP-side scheduling and dealbot's polling cadence. | [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts) |
```
